### DISCUSSION TITLE
A 'kind' which produces arbitrary outputs

### DISCUSSION THREAD

**Discussion by IcedQuinn** [xmake-io/xmake#1717] [Ideas]
Currently there is the `objects` kind which can create arbitrary outputs but it no longer can make use of the filename target fields. This new kind would basically be the same thing but still able to use the filename fields.

The use cases would be for example:
- Binding an epub ebook from a number of input files; the output would be a .epub (zip file)
- Making a GNU Info document for a project (which involves compiling texinfo to get the final .info file)
- Compile scripts and loose assets for a quest in to a single level file (for ex. rpg games.)

It looks like there is already a way to bring your own kinds but they are tied in to the toolchain system. That could probably also work if kinds (and the tools for them) can be defined in a project specific way (ex. so the project can just bring its own ebook toolchain code and use `set_kind(epub)` or something internally.)

**Comment by waruqi** [comment]
please use custom rules `add_rules("epub")` and set kind as `phony` or rewrite on_build or on_link

```lua
rule("epub")
    set_extensions("*.xxx")
    on_buildcmd_files(function (target, batchcmds, sourcefile, opt)
        -- handle file (optional)
    end)
    on_link(function (target)
         -- generate zip
    end)

target("test")
    add_rules("epub")
    add_files("src/*.xxx")
```

**Reply by IcedQuinn** [reply]
The target here doesn't have a set_kind call and will throw an error about not being able to link test with the C linker.

I know that set_kind('objects') will work with a custom rule but will set_kind('phony') still resolve dependencies? Usually phony targets in build systems explicitly don't have dependencies.

**Reply by waruqi** [reply]
you can use phony and you can also use target:set("kind", "xxx") to set kind in rule/on_load
